lib/ts/recipe/session/recipeImplementation.ts

```diff
@@ -323,48 +419,112 @@ export default function getRecipeInterface(
                     message: "Refresh token not found. Are you sending the refresh token in the request as a cookie?",
+                    payload: {
+                        clearTokens: false,
+                    },
                     type: STError.UNAUTHORISED,
                 });
             }
-            let antiCsrfToken = getAntiCsrfTokenFromHeaders(req);
-            let response = await SessionFunctions.refreshSession(
-                helpers,
-                inputRefreshToken,
-                antiCsrfToken,
-                getRidFromHeader(req) !== undefined
-            );
-            attachCreateOrRefreshSessionResponseToExpressRes(config, res, response);
-            logDebugMessage("refreshSession: Success!");
-            return new Session(
-                helpers,
-                response.accessToken.token,
-                response.session.handle,
-                response.session.userId,
-                response.session.userDataInJWT,
-                res
-            );
+
+            try {
+                let antiCsrfToken = getAntiCsrfTokenFromHeaders(req);
+                let response = await SessionFunctions.refreshSession(
+                    helpers,
+                    refreshToken,
+                    antiCsrfToken,
+                    getRidFromHeader(req) !== undefined,
+                    requestTransferMethod
+                );
+                logDebugMessage("refreshSession: Attaching refreshed session info as " + requestTransferMethod);
+
+                // We clear the tokens in all token transfer methods we are not going to overwrite
+                for (const transferMethod of availableTokenTransferMethods) {
+                    if (transferMethod !== requestTransferMethod && refreshTokens[transferMethod] !== undefined) {
+                        clearSession(config, res, transferMethod);
+                    }
+                }
+
+                attachCreateOrRefreshSessionResponseToExpressRes(config, res, response, requestTransferMethod);
+
+                logDebugMessage("refreshSession: Success!");
+                return new Session(
+                    helpers,
+                    response.accessToken.token,
+                    response.session.handle,
+                    response.session.userId,
+                    response.session.userDataInJWT,
+                    res,
+                    req,
+                    requestTransferMethod
+                );
+            } catch (err) {
+                if (
+                    (err.type === STError.UNAUTHORISED && err.payload.clearTokens) ||
+                    err.type === STError.TOKEN_THEFT_DETECTED
+                ) {
+                    logDebugMessage(
+                        "refreshSession: Clearing tokens because of UNAUTHORISED or TOKEN_THEFT_DETECTED response"
+                    );
+                    clearSession(config, res, requestTransferMethod);
+                }
```

Comment:
won't the errorhandler itself clear the cookies here from the session by catching these error types? Why do we need this check here as well?

Reply:
I haven't updated these during the merge (that introduced the errors clearing everything). I'll just remove these sections.